src/OrleansRuntime/Streams/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -0,0 +1,345 @@
+            return MaintainAndBalanceQueues(null);
+        }
+
+        public override IEnumerable<QueueId> GetMyQueues()
+        {
+            return this.myQueues.Select(queue => queue.QueueId);
+        }
+
+        private async Task MaintainAndBalanceQueues(object state)
+        {
+            CalculateResponsibility();
+            // step 1: renew existing leases 
+            await this.RenewLeases();
+            // step 2: if after renewing leases, myQueues count doesn't fall in [minimumResponsibility, maximumResponsibilty] range, act accordingly
+            if (this.myQueues.Count < this.minimumResponsibilty)
+            {
+                await this.AcquireLeasesToMeetMinResponsibility();
+            }
+            else if (this.myQueues.Count > this.maximumRespobsibility)
+            {
+                await this.ReleaseLeasesToMeetResponsibility();
+            }
+        }
+
+        private async Task ReleaseLeasesToMeetResponsibility()
+        {
+            var queueCountToRelease = this.myQueues.Count - this.maximumRespobsibility;
+            if (queueCountToRelease <= 0)
+                return;
+            var queuesToGiveUp = this.myQueues.GetRange(0, queueCountToRelease);
+            await this.leaseProvider.Release(ResourceCategory.Streaming, queuesToGiveUp.Select(queue => queue.AcquiredLease).ToArray());
+            //remove queuesToGiveUp from myQueue list after the balancer released the leases on them
+            this.myQueues.RemoveRange(0, queueCountToRelease);
+            this.logger.Info($"ReleaseLeasesToMeetResponsibility: released {queueCountToRelease} queues, current queue Count: {this.myQueues.Count}");
+        }
+
+        private Task AcquireLeaseToMeetMaxResponsibilty(object state)
+        {
+            return AcquireLeasesToMeetExpectation(this.maximumRespobsibility);
+        }
+
+        private Task AcquireLeasesToMeetMinResponsibility()
+        {
+            return AcquireLeasesToMeetExpectation(this.minimumResponsibilty);
+        }
+
+        private async Task AcquireLeasesToMeetExpectation(int expectedTotalLeaseCount)
+        {
+            int maxAttempts = 5;
+            int attempts = 0;
+            int leasesToAquire = expectedTotalLeaseCount - this.myQueues.Count;
+            this.logger.Info($"AcquireLeasesToMeetExpection : Try to acquire {leasesToAquire} queues");
+            while (attempts ++ <= maxAttempts && leasesToAquire > 0)
+            {
+                //select new queues to acquire
+                var expectedQueues = this.queueSelector.NextSelection(leasesToAquire, this.myQueues.Select(queue=>queue.QueueId)).ToList();
+                var leaseRequests = expectedQueues.Select(queue => new LeaseRequest() {
+                    ResourceKey = queue.ToString(),
+                    Duration = this.leaseLength
+                });
+                var results = await this.leaseProvider.Acquire(ResourceCategory.Streaming, leaseRequests.ToArray());
```

Comment:
maybe I'm wrong, but I would have expected this to include the name of the provider, possibly prefixed with the streaming category.

Otherwise there will be clashes between different stream providers configured with the same lease provider. Is that true?